storm-core/src/jvm/backtype/storm/windowing/WaterMarkEventGenerator.java

```diff
@@ -0,0 +1,110 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package backtype.storm.windowing;
+
+import backtype.storm.generated.GlobalStreamId;
+import backtype.storm.topology.FailedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Tracks tuples across input streams and periodically emits watermark events.
+ * Watermark event timestamp is the minimum of the latest tuple timestamps
+ * across all the input streams (minus the lag). Once a watermark event is emitted
+ * any tuple coming with an earlier timestamp can be considered as late events.
+ */
+public class WaterMarkEventGenerator<T> implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(WaterMarkEventGenerator.class);
+    private final WindowManager<T> windowManager;
+    private final int eventTsLag;
+    private final Set<GlobalStreamId> inputStreams;
+    private final Map<GlobalStreamId, Long> streamToTs;
+    private final ScheduledExecutorService executorService;
+    private final ScheduledFuture<?> executorFuture;
+    private long lastWaterMarkTs = 0;
+
+    public WaterMarkEventGenerator(WindowManager<T> windowManager, int interval,
+                                   int eventTsLag, Set<GlobalStreamId> inputStreams) {
+        this.windowManager = windowManager;
+        streamToTs = new ConcurrentHashMap<>();
+        executorService = Executors.newSingleThreadScheduledExecutor();
+        this.executorFuture = executorService.scheduleAtFixedRate(this, interval, interval, TimeUnit.MILLISECONDS);
+        this.eventTsLag = eventTsLag;
+        this.inputStreams = inputStreams;
+    }
+
+    public void track(GlobalStreamId stream, long ts) {
+        Long currentVal = streamToTs.get(stream);
+        if (currentVal == null || ts > currentVal) {
+            streamToTs.put(stream, ts);
+        }
+        checkFailures();
+    }
+
+    @Override
+    public void run() {
+        try {
+            long waterMarkTs = computeWaterMarkTs();
+            if (waterMarkTs > lastWaterMarkTs) {
+                this.windowManager.add(new WaterMarkEvent<T>(waterMarkTs - eventTsLag));
+                lastWaterMarkTs = waterMarkTs;
+            }
+        } catch (Throwable th) {
+            LOG.error("Failed while processing watermark event ", th);
+            throw th;
+        }
+    }
+
+    /**
+     * Computes the min ts across all streams.
+     */
+    private long computeWaterMarkTs() {
+        long ts = Long.MIN_VALUE;
+        // only if some data has arrived on each input stream
```

Comment:
The minimum of the latest event timestamps across all input streams (minus the lag) is considered as the watermark timestamp. This is so that if events from one of the streams is delayed more, we don't treat all events from that stream as late events.